File: src/main/java/ninekothecat/catconomy/defaultImplementations/CatMapDBDatabase.java
package ninekothecat.catconomy.defaultImplementations;

import ninekothecat.catconomy.Catconomy;
import ninekothecat.catconomy.interfaces.IDatabase;
import org.mapdb.DB;
import org.mapdb.DBMaker;

import java.io.File;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentMap;

@SuppressWarnings("unchecked")
public class CatMapDBDatabase implements IDatabase {
    private final DBMaker.Maker fileDB;

    public CatMapDBDatabase() {
        File dbFile = new File(Catconomy.getProvidingPlugin(Catconomy.class).getDataFolder(), "accounts.db");
        fileDB = DBMaker.fileDB(dbFile);
    }

    @Override
    public boolean userExists(UUID user) {
        DB db = fileDB.make();
        ConcurrentMap<UUID, Double> user_map = (ConcurrentMap<UUID, Double>) db.hashMap("accounts").createOrOpen();
        boolean exists = user_map.containsKey(user);
        db.close();
        return exists;
    }

    @Override
    public void removeUser(UUID user) {
        DB db = fileDB.make();
        ConcurrentMap<UUID, Double> user_map = (ConcurrentMap<UUID, Double>) db.hashMap("accounts").createOrOpen();
        user_map.remove(user);
        db.commit();
        db.close();
    }

    @Override
    public void setUsersBalance(Map<UUID, Double> userBalances) {
        DB db = fileDB.make();
        ConcurrentMap<UUID, Double> user_map = (ConcurrentMap<UUID, Double>) db.hashMap("accounts").createOrOpen();

        for (UUID user : userBalances.keySet()) {
            user_map.put(user, userBalances.get(user));
        }
        db.commit();
        db.close();
    }

    @Override
    public Map<UUID, Double> getUsersBalance(Collection<UUID> users) {
        DB db = fileDB.make();
        ConcurrentMap<UUID, Double> user_map = (ConcurrentMap<UUID, Double>) db.hashMap("accounts").createOrOpen();
        Map<UUID, Double> gottenUsers = new HashMap<>();
        for (UUID user : users) {
            gottenUsers.put(user, user_map.get(user));
        }
        db.close();
        return gottenUsers;
    }
}

File: src/main/kotlin/ninekothecat/catconomy/defaultImplementations/database/CatMapDBDatabase.kt
package ninekothecat.catconomy.defaultImplementations.database

import ninekothecat.catconomy.Catconomy
import ninekothecat.catconomy.interfaces.IDatabase
import org.bukkit.plugin.java.JavaPlugin
import org.mapdb.DB
import org.mapdb.DBMaker.fileDB
import java.io.File
import java.util.*
import java.util.concurrent.ConcurrentMap

@Suppress("UNCHECKED_CAST")
class CatMapDBDatabase : IDatabase {
    companion object {
        var db : DB? = null
    }


    override fun userExists(user: UUID): Boolean {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double>
        return userMap.containsKey(user)
    }

    override fun removeUser(user: UUID) {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double>
        userMap.remove(user)
        db!!.commit()
    }

    override fun setUserBalance(user: UUID, balance: Double) {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double>
        userMap[user] = balance
        db!!.commit()
    }

    override fun getUserBalance(user: UUID): Double {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double>
        val amount = userMap.getOrDefault(user, Catconomy.startingAmount)
        return amount
    }

    override fun setUsersBalance(userBalances: MutableMap<UUID, Double>) {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double?>
        for (user in userBalances.keys) {
            userMap[user] = userBalances[user]
        }
        db!!.commit()
    }

    override fun getUsersBalance(users: Collection<UUID>): Map<UUID, Double?> {
        val userMap = db?.hashMap("accounts")?.createOrOpen() as ConcurrentMap<UUID, Double>
        val gottenUsers: MutableMap<UUID, Double?> = HashMap()
        for (user in users) {
            gottenUsers[user] = userMap[user]
        }
        return gottenUsers
    }

    init {
        val dbFile = File(JavaPlugin.getProvidingPlugin(Catconomy::class.java).dataFolder, "accounts.db")
        if (db == null){
            db = fileDB(dbFile).make()
        }
    }

}